zebra-network/src/config.rs

```diff
@@ -176,3 +184,86 @@ impl Default for Config {
         }
     }
 }
+
+impl<'de> Deserialize<'de> for Config {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(deny_unknown_fields, default)]
+        struct DConfig {
+            listen_addr: String,
+            network: Network,
+            initial_mainnet_peers: HashSet<String>,
+            initial_testnet_peers: HashSet<String>,
+            peerset_initial_target_size: usize,
+            #[serde(alias = "new_peer_interval")]
+            crawl_new_peer_interval: Duration,
+        }
+
+        impl Default for DConfig {
+            fn default() -> Self {
+                let config = Config::default();
+                Self {
+                    listen_addr: config.listen_addr.to_string(),
+                    network: config.network,
+                    initial_mainnet_peers: config.initial_mainnet_peers,
+                    initial_testnet_peers: config.initial_testnet_peers,
+                    peerset_initial_target_size: config.peerset_initial_target_size,
+                    crawl_new_peer_interval: config.crawl_new_peer_interval,
+                }
+            }
+        }
+
+        let config = DConfig::deserialize(deserializer)?;
+        let listen_addr = match config.listen_addr.parse::<SocketAddr>() {
+            Ok(socket) => Ok(socket),
+            Err(_) => match config.listen_addr.parse::<IpAddr>() {
```

Comment:
This refactor will also help us do another ticket, so let's mention it here:
```suggestion
        /// TODO: perform listener DNS lookups asynchronously with a timeout (#1631)
        let listen_addr = match config.listen_addr.parse::<SocketAddr>() {
            Ok(socket) => Ok(socket),
            Err(_) => match config.listen_addr.parse::<IpAddr>() {
```